pipeline/src/lib.rs

```diff
@@ -46,21 +49,54 @@ pub fn access_element<T: FieldElement>(
 }
 
 #[allow(clippy::print_stdout)]
-pub fn inputs_to_query_callback<T: FieldElement>(inputs: Vec<T>) -> impl QueryCallback<T> {
+pub fn serde_data_to_query_callback<T: FieldElement, S: serde::Serialize + Send + Sync>(
+    channel: u32,
+    data: &S,
+) -> impl QueryCallback<T> {
+    let bytes = serde_json::to_vec(&data).unwrap();
+    move |query: &str| -> Result<Option<T>, String> {
+        match &parse_query(query)?[..] {
+            ["\"data\"", index, what] => {
+                let what = what
+                    .parse::<u32>()
+                    .map_err(|e| format!("Error parsing what: {e})"))?;
+
+                if what != channel {
+                    return Ok(None);
+                }
+
+                let index = index
+                    .parse::<usize>()
+                    .map_err(|e| format!("Error parsing index: {e})"))?;
+
+                // query index 0 means the length
+                Ok(Some(match index {
+                    0 => (bytes.len() as u64).into(),
+                    index => (bytes[index - 1] as u64).into(),
+                }))
+            }
+            k => Err(format!("Unsupported query: {}", k.iter().format(", "))),
+        }
+    }
+}
+#[allow(clippy::print_stdout)]
+pub fn inputs_to_query_callback<T: FieldElement>(
+    inputs: HashMap<u32, Vec<T>>,
+) -> impl QueryCallback<T> {
     move |query: &str| -> Result<Option<T>, String> {
         // TODO In the future, when match statements need to be exhaustive,
         // This function probably gets an Option as argument and it should
         // answer None by Ok(None).
 
         match &parse_query(query)?[..] {
-            ["\"input\"", index] => access_element("prover inputs", &inputs, index),
+            ["\"input\"", index] => access_element("prover inputs", &inputs[&0], index),
             ["\"data\"", index, what] => {
                 let what = what
-                    .parse::<usize>()
+                    .parse::<u32>()
                     .map_err(|e| format!("Error parsing what: {e})"))?;
                 assert_eq!(what, 0);
```

Comment:
```suggestion
                assert_eq!(what, 0);
```
I guess this can be removed now?

Reply:
I thought this should stay because this function is used to create callbacks for the input vector case. We use another function to make callbacks for serde data. Or do you see it differently?

Reply:
I think either this function should accept a `inputs: Vec<T>` as before or truly support several arrays (which it could by removing the `assert`). I don't think it makes too much sense to accept a Map at compile time and then fail at runtime.

Reply:
Right